### PR TITLE
Gamebanana watcher

### LINKED ISSUE #7: GameBanana watcher
Check [GameBanana](https://gamebanana.com) for new CS:S and CS:GO bhop/kz maps, if there is a new submission send a message to a specific channel.

gamebanana

### PR DESCRIPTION
**Pull request by log-ical**
From what I've been able to see it works, however, I may have overlooked something. There are several things that had to be taken into account and there may be a few things worthy of concern here:
   - Ensuring entries don't send multiple times
   In regards to ensuring entries don't send multiple times, I used a method of pushing an object (containing the name, uploaded date, and updated date) into an array for each version of the game. The setback with this method, however, is that if the bot ever restarts that array will be cleared and the most recent CS:S and CS:GO map will be sent. 
   - Differentiating between an update and a new post
   For this, I just compare the updated date previously stored in the map object to the new updated date retrieved from the API. The only problem with this is I'm not sure if the `_tsDateUpdated` from the API is the correct number I should be using for the upload date. When comparing the API data to the entries already in the SJ discord sometimes the `_tsDateUpdated` value would match what I was seeing, but sometimes it wouldn't. For example:
image
This is what it shows for this map on the API, but in SJ it shows this.
image
The only number that is changed for this particular example is `_tsDateModified`, however, it can't be this value, because this changes any time something such as the description of a map is changed. Changes such as those are not reflected in the bot already present in the SJ discord. 
   I'm just going to assume in this instance, it was the bot that is already on SJ which made the mistake. Looking at the GameBanana page for the map it doesn't appear like anything was actually updated. 

tl;dr 
1.) If the bot ever restarts it will send the most recent map from CS:S and CS:GO respectively. 
2.) I have not been able to test and there *may* be an issue with how I handle if a map is updated or not, however, from what I can see everything should work fine.

Let me know if you know of a better way to do all of this.

In any case... 🙂 
This is what I was able to recreate: 

image

### CODE REVIEW

File: src/events/gamebanana_watcher.js [hunk 1]
```diff
@@ -0,0 +1,144 @@
+require('dotenv').config();
```
Comment (ecsr) [inline src/events/gamebanana_watcher.js]: The .env file has already been loaded in index.js so this is not needed.

File: src/events/gamebanana_watcher.js [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        const csgoUrl =
+            'https://gamebanana.com/mods/cats/5422?api=SubmissionsListModule';
+        let guild = client.guilds.resolve(process.env.GUILD_ID);
+        guild.channels
+            .fetch(process.env.NEW_MAPS_CHANNEL_ID)
+            .then((GuildChannel) => {
+                GuildChannel.fetch().then((channel) => {
+                    setInterval(async () => {
+                        fetch(cssUrl)
+                            .then((data) => data.text())
```
Comment (ecsr) [inline src/events/gamebanana_watcher.js]:
```suggestion
                            .then((data) => data.json())
```
To access the JSON data directly.

File: src/events/gamebanana_watcher.js [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+                            .then((body) => {
+                                body = JSON.parse(body);
```
Comment (ecsr) [inline src/events/gamebanana_watcher.js]: This can then be removed if the above is changed.

File: src/events/gamebanana_watcher.js [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+module.exports = {
+    name: 'ready',
+    execute: async (client) => {
+        let cssMaps = [];
+        let csgoMaps = [];
+        const cssUrl =
+            'https://gamebanana.com/mods/cats/5568?api=SubmissionsListModule';
+        const csgoUrl =
+            'https://gamebanana.com/mods/cats/5422?api=SubmissionsListModule';
```
Comment (ecsr) [inline src/events/gamebanana_watcher.js]: Maybe use the official GameBanana API instead (even though its horrible) otherwise we might get rate limited. [Here's an example](https://gamebanana.com/apiv8/Mod/ByCategory?_csvProperties=_idRow,_tsDateAdded,_tsDateUpdated,_sName,_aSubmitter&_aCategoryRowIds[]=5568&_sOrderBy=_tsDateAdded,DESC&_nPerpage=10) changing `&_aCategoryRowIds[]=5568` to `&_aCategoryRowIds[]=5422` will get you CSGO maps instead.

File: src/events/gamebanana_watcher.js [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+                                        {
+                                            name: 'Download',
+                                            value: `${mapDL} `,
+                                        },
+                                    ])
+                                    .setImage(mapPreviewImage);
+                                channel.send({
+                                    embeds: [embed],
+                                });
+                            });
```
Comment (ecsr) [inline src/events/gamebanana_watcher.js]: This part is pretty much identical to the code for checking CSS submissions and it can probably be made into its own function that handles both cases.

File: src/events/gamebanana_watcher.js [hunk 6]
```diff
@@ -0,0 +1,144 @@
+require('dotenv').config();
+const { MessageEmbed } = require('discord.js');
+const fetch = require('node-fetch');
+
+module.exports = {
+    name: 'ready',
+    execute: async (client) => {
+        let cssMaps = [];
+        let csgoMaps = [];
```
Comment (ecsr) [inline src/events/gamebanana_watcher.js]: No need to create 2 separate arrays here instead use 1 array and store mod ID to avoid duplicates. Also maybe we should consider storing maps in a database instead? A database will be needed for a few other things in the bot so we will have to set one up anyway.

**Comment by log-ical** [comment]
Updated to match suggestions. I do agree moving to a database would probably be the best in the long term, it will fix this issue as well:
> 1.) If the bot ever restarts it will send the most recent map from CS:S and CS:GO respectively.

However, I didn't add support for it in the most recent commit. Do we want to do that all now?

**Comment by ecsr** [comment]
> However, I didn't add support for it in the most recent commit. Do we want to do that all now?

Database support (https://github.com/sourcejump/discord-bot/pull/10) is in progress so lets wait until that is done before merging this.

**Comment by ecsr** [comment]
See https://github.com/sourcejump/discord-bot/pull/11 and https://typeorm.io for database integration.

**Comment by log-ical** [comment]
I moved everything over to use the database.
It creates a gamebanana_maps table and it looks like so: 
image

There are 7 columns
- `id` - Autoincremented id, keeps the map added in order.
- `modID`
- `mapName`
- `mapAuthor`
- `mapPreviewImage`
- `dateAdded`
- `dateLastUpdated`

We store the map every time it is not already in the table. If the map already exists in the table and the update time does not match with what was previously stored then we update the `dateLastUpdated` in the table and change the embed to say "Map Updated". 

image
image
image

File: src/database/entity/gamebanana_maps.js [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+        },
+        mapName: {
+            type: 'varchar',
+        },
+        mapAuthor: {
+            type: 'varchar',
+        },
+        mapPreviewImage: {
+            type: 'varchar',
+        },
```
Comment (ecsr) [inline src/database/entity/gamebanana_maps.js]: Storing image URL in the database is not really needed as we can recreate that using the mod ID.

**Comment by log-ical** [comment]
changed

**Review by ecsr — APPROVED** [review]
